cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,831 @@
+  using layout_type  = LayoutPolicy;
+  using mapping_type = typename layout_type::template mapping<extents_type>;
+  using element_type = ElementType;
+
+  using value_type      = std::remove_cv_t<element_type>;
+  using index_type      = typename extents_type::index_type;
+  using difference_type = std::ptrdiff_t;
+  using rank_type       = typename extents_type::rank_type;
+
+  using container_policy_type   = ContainerPolicy;
+  using accessor_policy_variant = typename ContainerPolicy::accessor_policy_variant;
+
+  template <raft::memory_type MemType>
+  using accessor_policy = alternate_from_mem_type<MemType, accessor_policy_variant>;
+
+  using container_type_variant = typename container_policy_type::container_type_variant;
+
+  template <raft::memory_type MemType>
+  using container_type = typename container_policy_type::template container_type<MemType>;
+
+  template <memory_type MemType>
+  using owning_type = mdarray<element_type,
+                              extents_type,
+                              layout_type,
+                              typename container_policy_type::template container_policy<MemType>>;
+  // We use the static cast here to ensure that the memory types appear in the
+  // order expected for retrieving the correct variant alternative based on
+  // memory type. Even if the memory types are re-arranged in the enum and
+  // assigned different values, the logic should remain correct.
+  using owning_type_variant = std::variant<owning_type<static_cast<memory_type>(0)>,
+                                           owning_type<static_cast<memory_type>(1)>,
+                                           owning_type<static_cast<memory_type>(2)>,
+                                           owning_type<static_cast<memory_type>(3)>>;
+
+  template <memory_type MemType>
+  using view_type = typename owning_type<MemType>::view_type;
+
+  using view_type_variant = std::variant<view_type<static_cast<memory_type>(0)>,
+                                         view_type<static_cast<memory_type>(1)>,
+                                         view_type<static_cast<memory_type>(2)>,
+                                         view_type<static_cast<memory_type>(3)>>;
+
+  template <memory_type MemType>
+  using const_view_type         = typename owning_type<MemType>::const_view_type;
+  using const_view_type_variant = std::variant<const_view_type<static_cast<memory_type>(0)>,
+                                               const_view_type<static_cast<memory_type>(1)>,
+                                               const_view_type<static_cast<memory_type>(2)>,
+                                               const_view_type<static_cast<memory_type>(3)>>;
+
+  using storage_type_variant = concatenated_variant_t<view_type_variant, owning_type_variant>;
+
+  template <memory_type MemType, bool is_owning>
+  using storage_type =
+    std::variant_alternative_t<std::size_t{is_owning} * std::variant_size_v<view_type_variant> +
```

Comment:
Can you write a small example in commented code here explaining this arithmetic?